### DISCUSSION TITLE
Error 1030: Got error 168 - 'Unknown (generic) error from engine' from storage engine

### DISCUSSION THREAD

**Discussion by Nas-au** [authelia/authelia#3560] [Q&A] [answered]
time="2022-06-20T22:17:23Z" level=info msg="Authelia v4.35.6 is starting"
time="2022-06-20T22:17:23Z" level=info msg="Log severity set to debug"
time="2022-06-20T22:17:23Z" level=info msg="Storage schema is being checked for updates"
time="2022-06-20T22:17:24Z" level=info msg="Storage schema migration from 0 to 5 is being attempted"
time="2022-06-20T22:17:24Z" level=error msg="Failure running the storage provider startup check: error during schema migrate: migration rollback complete. rollback caused by: schema migration 1 (Initial Schema) failed: Error 1030: Got error 168 - 'Unknown (generic) error from engine' from storage engine" stack="github.com/authelia/authelia/v4/internal/commands/root.go:94 doStartupChecks\ngithub.com/authelia/authelia/v4/internal/commands/root.go:78 cmdRootRun\ngithub.com/spf13/cobra@v1.4.0/command.go:860                 (*Command).execute\ngithub.com/spf13/cobra@v1.4.0/command.go:974                 (*Command).ExecuteC\ngithub.com/spf13/cobra@v1.4.0/command.go:902                 (*Command).Execute\ngithub.com/authelia/authelia/v4/cmd/authelia/main.go:10      main\nruntime/proc.go:250                                          main\nruntime/asm_amd64.s:1571                                     goexit"
time="2022-06-20T22:17:24Z" level=debug msg="LDAP Supported OIDs. Control Types: 2.16.840.1.113730.3.4.9, 1.2.840.113556.1.4.473, 1.3.6.1.4.1.4203.1.9.1.1, 2.16.840.1.113730.3.4.5, 2.16.840.1.113730.3.4.4, 1.3.6.1.4.1.42.2.27.8.5.1, 2.16.840.1.113730.3.4.18, 2.16.840.1.113730.3.4.2, 1.3.6.1.4.1.4203.1.10.1, 1.3.6.1.1.22, 1.2.840.113556.1.4.319, 1.2.826.0.1.3344810.2.3, 1.3.6.1.1.13.2, 1.3.6.1.1.13.1, 1.3.6.1.1.12. Extensions: 1.3.6.1.4.1.1466.20037, 1.3.6.1.4.1.4203.1.11.1, 1.3.6.1.4.1.4203.1.11.3, 1.3.6.1.1.8"
time="2022-06-20T22:17:24Z" level=error msg="Your LDAP Server supports TLS but you don't appear to be utilizing it. We stronglyrecommend enabling the StartTLS option or using the scheme 'ldaps://' to secure connections with yourLDAP Server." stack="github.com/authelia/authelia/v4/internal/commands/root.go:136 doStartupCheck\ngithub.com/authelia/authelia/v4/internal/commands/root.go:99  doStartupChecks\ngithub.com/authelia/authelia/v4/internal/commands/root.go:78  cmdRootRun\ngithub.com/spf13/cobra@v1.4.0/command.go:860                  (*Command).execute\ngithub.com/spf13/cobra@v1.4.0/command.go:974                  (*Command).ExecuteC\ngithub.com/spf13/cobra@v1.4.0/command.go:902                  (*Command).Execute\ngithub.com/authelia/authelia/v4/cmd/authelia/main.go:10       main\nruntime/proc.go:250                                           main\nruntime/asm_amd64.s:1571                                      goexit"
time="2022-06-20T22:17:24Z" level=fatal msg="The following providers had fatal failures during startup: storage" stack="github.com/authelia/authelia/v4/internal/commands/root.go:122 doStartupChecks\ngithub.com/authelia/authelia/v4/internal/commands/root.go:78  cmdRootRun\ngithub.com/spf13/cobra@v1.4.0/command.go:860                  (*Command).execute\ngithub.com/spf13/cobra@v1.4.0/command.go:974                  (*Command).ExecuteC\ngithub.com/spf13/cobra@v1.4.0/command.go:902                  (*Command).Execute\ngithub.com/authelia/authelia/v4/cmd/authelia/main.go:10       main\nruntime/proc.go:250                                           main\nruntime/asm_amd64.s:1571                                      goexit"

my config

authentication_backend:
  ldap:
    implementation: custom
    url: ldap://xxxxxxxxxx
    start_tls: false
    base_dn: dc=ldap,dc=xxxx,dc=xxxx
    username_attribute: uid
    group_name_attribute: cn
    mail_attribute: mail
    display_name_attribute: gecos
    additional_users_dn: cn=users
    additional_groups_dn: cn=groups
    users_filter: (&(|({username_attribute}={input})({mail_attribute}={input}))(objectClass=person))
    groups_filter: (member={dn})
    user: uid=root,cn=users,dc=ldap,dc=xxxx,dc=xxxx
    password: xxxxxxx

**Comment by james-d-elliott** [comment]
That's an error with your storage provider as clearly stated by `The following providers had fatal failures during startup: storage`.

The actual error is this:

```text
Failure running the storage provider startup check: error during schema migrate: migration rollback complete. rollback caused by: schema migration 1 (Initial Schema) failed: Error 1030: Got error 168 - 'Unknown (generic) error from engine' from storage engine
```

We test against all supported databases, and there is not enough information here to replicate this issue. Please provide information that allows replicating this issue.

**Comment by james-d-elliott** [comment, marked as answer]
As this article exists on the error I'm converting this to a Q&A. See here: https://ao.ms/solution-error-1030-hy000-got-error-168-from-storage-engine/

**Reply by james-d-elliott** [reply]
@Nas-au did this solve your issue? If so can you mark it as the answer.

**Reply by Nas-au** [reply]
Mysql did not work i changed the storage to local "sqlite3" and it worked fine.
thank you

**Reply by james-d-elliott** [reply]
No worries, according to several docs this error is returned by MySQL when it can't write to disk (due to permissions or error).

**Reply by Nas-au** [reply]
Thank you for pointing me to the right direction, your right it was a permission error.

it works fine now

Thanks again.

**Comment by kyle-belle** [comment]
I am also getting this error. simple by adding `@db.Text` to a string field 
```md
- photoUrl      String?
+ photoUrl      String?       @db.Text
```

```
Error: Got error 168 - 'Unknown (generic) error from engine' from storage engine
   0: sql_schema_connector::apply_migration::migration_step
           with step=AlterTable(AlterTable { table_ids: MigrationPair { previous: TableId(4), next: TableId(0) }, changes: [AlterColumn(AlterColumn { column_id: MigrationPair { previous: TableColumnId(49), next: TableColumnId(9) }, changes: ColumnChanges { type_change: Some(SafeCast), changes: BitFlags<ColumnChange>(0b100, TypeChanged) }, type_change: Some(SafeCast) })] })
             at schema-engine/connectors/sql-schema-connector/src/apply_migration.rs:21
   1: sql_schema_connector::apply_migration::apply_migration
             at schema-engine/connectors/sql-schema-connector/src/apply_migration.rs:10
   2: schema_core::state::SchemaPush
             at schema-engine/core/src/state.rs:436

error Command failed with exit code 1.
```

**Reply by james-d-elliott** [reply]
This is not a error related to Authelia at all seems like a rust project of some kind? Also the underlying error is a error with the database file permissions.

**Reply by kyle-belle** [reply]
oh damn. didn't even realize i wasn't in a prisma specific repo. my bad

**Reply by james-d-elliott** [reply]
No big deal, but the underlying cause for this was a file permission error on the database data files/directory from memory.

**Reply by kyle-belle** [reply]
I removed the docker container containing my db and started it up again and now it works so yeah most likely was some kind of permission issue
